src/Interpreters/ThreadStatusExt.cpp

```diff
@@ -337,7 +337,8 @@ void ThreadStatus::detachQuery(bool exit_if_already_detached, bool thread_exits)
     /// (this is in case someone uses system thread but did not call getProfileEventsCountersAndMemoryForThreads())
     {
         std::lock_guard guard(thread_group->mutex);
-        auto stats = std::move(thread_group->finished_threads_counters_memory);
+        thread_group->finished_threads_counters_memory.resize(0);
+        thread_group->finished_threads_counters_memory.shrink_to_fit();
```

Comment:
I'm not sure if these two lines are actually more readable than that one, but equivalent. But I feel like an idea of a code author was to avoid destruction under lock (and the current code is not doing this due to destruction order). Maybe you could improve this point here instead. First, move under lock, then destruct this heavy vector w/o locking the whole thread_group?

Reply:
Thanks for sharing your insight! I will try to revise the patch accordingly.

Reply:
Sergei, I've tried to revise the patch and evaluate the performance impact. The local performance test shows that this change could bring performance gain to some workloads, however, leads to slight regression in others. I'm working on the performance analysis and will update my findings once ready.

Thanks again!